### PR TITLE
Add GCS authentication for service accounts

### PR DESCRIPTION
**Pull request by b-chu**
## Description of changes:
Allow authentication with GCS for service accounts

## Issue #, if available:
STR-69

## Merge Checklist:
_Put an `x` without space in the boxes that apply. If you are unsure about any checklist, please don't hesitate to ask. We are here to help! This is simply a reminder of what we are going to look for before merging your pull request._

### General
- [x] I have read the [contributor guidelines](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md)
- [ ] This is a documentation change or typo fix. If so, skip the rest of this checklist.
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the MosaicML team.
- [x] I have updated any necessary documentation, including [README](https://github.com/mosaicml/streaming/blob/main/README.md) and [API docs](https://github.com/mosaicml/streaming/tree/main/docs) (if appropriate).

### Tests
- [x] I ran `pre-commit` on my change. (check out the `pre-commit` section of [prerequisites](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites))
- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [x] I ran the tests locally to make sure it pass. (check out [testing](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#running-tests))
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes.

### CODE REVIEW

File: docs/source/how_to_guides/configure_cloud_storage_credentials.md [hunk 1]
```diff
@@ -116,6 +116,19 @@ export GCS_SECRET='wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY'
 ```
 ````
 
+Streaming dataset also supports [GCP service account credentials](https://cloud.google.com/iam/docs/service-account-overview). Users must set their GCP `account credentials` in the run environment.
```
Comment (karan6181) [inline docs/source/how_to_guides/configure_cloud_storage_credentials.md:119]: Can you please create a two sections, one for `GCP Service Account Credentials` and another for `GCP User Auth Credentials Mounted as Environment Variables` similar to [this](https://docs.mosaicml.com/projects/mcli/en/latest/resources/secrets/gcp.html) doc. Thanks!

File: streaming/base/storage/upload.py [hunk 2]
```diff
@@ -32,14 +37,21 @@
...
     @classmethod
-    def get(cls,
-            out: Union[str, Tuple[str, str]],
-            keep_local: bool = False,
-            progress_bar: bool = False) -> Any:
+    def get(
+        cls,
+        out: Union[str, Tuple[str, str]],
+        keep_local: bool = False,
+        progress_bar: bool = False,
```
Comment (karan6181) [inline streaming/base/storage/upload.py:53]: Any specific reason why you added trailing `,` at the end ?

File: streaming/base/storage/upload.py [hunk 3]
```diff
@@ -85,17 +98,20 @@ def _validate(self, out: Union[str, Tuple[str, str]]) -> None:
...
 
-    def __init__(self,
-                 out: Union[str, Tuple[str, str]],
-                 keep_local: bool = False,
-                 progress_bar: bool = False) -> None:
+    def __init__(
+        self,
+        out: Union[str, Tuple[str, str]],
+        keep_local: bool = False,
+        progress_bar: bool = False,
```
Comment (karan6181) [inline streaming/base/storage/upload.py:113]: Same here and other places.

File: streaming/base/storage/upload.py [hunk 4]
```diff
@@ -255,49 +276,80 @@ class GCSUploader(CloudUploader):
...
+                    desc=f'Uploading to {remote_filename}',
+                    disable=(not self.progress_bar),
+            ) as pbar:
+                self.gcs_client.upload_file(
+                    local_filename,
+                    obj.netloc,
+                    obj.path.lstrip('/'),
+                    Callback=lambda bytes_transferred: pbar.update(bytes_transferred),
+                )
+
```
Comment (karan6181) [inline streaming/base/storage/upload.py:343]: extra space

File: streaming/base/storage/upload.py [hunk 5]
```diff
@@ -336,13 +394,16 @@ class OCIUploader(CloudUploader):
...
 
-    def __init__(self,
-                 out: Union[str, Tuple[str, str]],
-                 keep_local: bool = False,
-                 progress_bar: bool = False) -> None:
+    def __init__(
+        self,
+        out: Union[str, Tuple[str, str]],
+        keep_local: bool = False,
+        progress_bar: bool = False,
```
Comment (karan6181) [inline streaming/base/storage/upload.py:401]: Do you mind reverting the trailing comma `,` changes as this is not related to this PR? Thanks!

File: tests/test_upload.py [hunk 6]
```diff
@@ -74,6 +85,7 @@ def test_delete_local_file(self, local_remote_dir: Tuple[str, str]):
     @pytest.mark.parametrize('out', ['s3://bucket/dir', 'gs://bucket/dir'])
     def test_check_bucket_exists_exception(self, out: str):
         import botocore
+
```
Comment (karan6181) [inline tests/test_upload.py:88]: extra line

File: docs/source/how_to_guides/configure_cloud_storage_credentials.md [hunk 7]
```diff
@@ -93,11 +93,28 @@ export S3_ENDPOINT_URL='https://<accountid>.r2.cloudflarestorage.com'
 
 ## Google Cloud Storage
 
-### MosaicML platform
+### GCP Service Account Credentials
+
+#### MosaicML platform
```
Comment (karan6181) [inline docs/source/how_to_guides/configure_cloud_storage_credentials.md:98]: I think it would be better to move the `MosaicML platform` above `GCP Service Account Credentials` ? since it will be same for HMAC and service level authenticaltion.

File: streaming/base/storage/download.py [hunk 8]
```diff
@@ -158,15 +158,33 @@ def download_from_gcs(remote: str, local: str) -> None:
...
     if obj.scheme != 'gs':
         raise ValueError(
             f'Expected obj.scheme to be `gs`, instead, got {obj.scheme} for remote={remote}')
 
+    if 'GOOGLE_APPLICATION_CREDENTIALS' in os.environ:
+        _gcs_with_service_account(local, obj)
+    elif 'GCS_KEY' in os.environ and 'GCS_SECRET' in os.environ:
+        _gcs_with_hmac(remote, local, obj)
+    else:
+        raise ValueError(('Either GOOGLE_APPLICATION_CREDENTIALS needs to be set for'
```
Comment (karan6181) [inline streaming/base/storage/download.py:171]: Extra `(` and `)` ?

File: streaming/base/storage/download.py [hunk 9]
```diff
@@ -190,6 +208,22 @@ def download_from_gcs(remote: str, local: str) -> None:
...
+        local (str): Local path (local filesystem).
+        obj (ParseResult): ParseResult object of remote path (GCS).
+    """
+    from google.cloud.storage import Blob, Bucket, Client
+
+    service_account_path = os.environ['GOOGLE_APPLICATION_CREDENTIALS']
+    gcs_client = Client.from_service_account_json(service_account_path)
+
+    blob = Blob(obj.path.lstrip('/'), Bucket(gcs_client, obj.netloc))
+    blob.download_to_filename(local)
```
Comment (karan6181) [inline streaming/base/storage/download.py:224]: Can you check the in-progress filename by google service account ? For example, if rank 0 is downloading an `index.json` file with let say 64MB and other ranks wait for file `index.json` to exists, then, if GCS downloads it as `index.json` with split chunks, meaning, the file is not downloaded fully and other ranks thinks that it is downloaded fully and starts loading it and this would cause JSONDecoder error. However, if GCS downloads it as some temp extension such as `index.json.tmp` or something similar and then later renames it to `index.json` then we are good here.

File: streaming/base/storage/upload.py [hunk 10]
```diff
@@ -32,6 +37,11 @@
 }
 
 
+class Authentication(Enum):
```
Comment (karan6181) [inline streaming/base/storage/upload.py:40]: Can we rename it to `_GCSAuthentication` ? `_` meaning private class

File: streaming/base/storage/upload.py [hunk 11]
```diff
@@ -60,7 +70,8 @@ def get(cls,
             CloudUploader: An instance of sub-class.
         """
         cls._validate(cls, out)
-        obj = urllib.parse.urlparse(out) if isinstance(out, str) else urllib.parse.urlparse(out[1])
+        obj = (urllib.parse.urlparse(out)
```
Comment (karan6181) [inline streaming/base/storage/upload.py:73]: Do we really need `(` before `urllib.parse` ?

File: streaming/base/storage/upload.py [hunk 12]
```diff
@@ -85,8 +96,9 @@ def _validate(self, out: Union[str, Tuple[str, str]]) -> None:
         else:
             if len(out) != 2:
                 raise ValueError(''.join([
-                    f'Invalid `out` argument. It is either a string of local/remote directory ',
-                    'or a list of two strings with [local, remote].'
+                    (f'Invalid `out` argument. It is either a string'
```
Comment (karan6181) [inline streaming/base/storage/upload.py:99]: Any reason for this change ?

File: streaming/base/storage/upload.py [hunk 13]
```diff
@@ -183,6 +195,7 @@ def __init__(self,
 
         import boto3
         from botocore.config import Config
+
```
Comment (karan6181) [inline streaming/base/storage/upload.py:198]: Extra space ?

File: streaming/base/storage/upload.py [hunk 14]
```diff
@@ -261,19 +274,34 @@ def __init__(self,
...
+            self.gcs_client = session.client(
+                's3',
+                region_name='auto',
+                endpoint_url='https://storage.googleapis.com',
+                aws_access_key_id=os.environ['GCS_KEY'],
+                aws_secret_access_key=os.environ['GCS_SECRET'],
+            )
+            self.authentication = Authentication.HMAC
+        else:
+            raise ValueError(('Either GOOGLE_APPLICATION_CREDENTIALS needs to be set for'
```
Comment (karan6181) [inline streaming/base/storage/upload.py:298]: Extra `(` and `)` ?

File: streaming/base/storage/upload.py [hunk 15]
```diff
@@ -471,7 +515,7 @@ def check_bucket_exists(self, remote: str):
             error: Bucket does not exist.
         """
         bucket_name = urllib.parse.urlparse(remote).netloc
-        if self.azure_service.get_container_client(container=bucket_name).exists() is False:
+        if (self.azure_service.get_container_client(container=bucket_name).exists() is False):
```
Comment (karan6181) [inline streaming/base/storage/upload.py:518]: Any reason for this change ? Extra `(` and `)` ?

File: streaming/base/storage/upload.py (hunk 14 above)
Comment (b-chu) [reply streaming/base/storage/upload.py:298]: If I don't have this pylance complains about implicit string concatenation.

File: streaming/base/storage/upload.py (hunk 12 above)
Comment (b-chu) [reply streaming/base/storage/upload.py:99]: Changed it again. Not sure why we needed f-string or concatenation using ''.join().

File: streaming/base/storage/upload.py (hunk 14 above)
Comment (b-chu) [reply streaming/base/storage/upload.py:298]: Changed it to be a single string so extra parentheses aren't needed, but I think the previous version looked more readable.

File: streaming/base/storage/upload.py (hunk 10 above)
Comment (b-chu) [reply streaming/base/storage/upload.py:40]: I think it should be kept public? If the authentication type needs to be checked by a caller, they would use self.authentication and compare to GCSAuthentication.{something} like how it's being tested.

File: streaming/base/storage/download.py (hunk 8 above)
Comment (b-chu) [reply streaming/base/storage/download.py:171]: Commented in upload.py

File: streaming/base/storage/download.py (hunk 9 above)
Comment (b-chu) [reply streaming/base/storage/download.py:224]: It seems like it opens a file object which it writes into but the file for a file object doesn't show up in the file system until after it's done writing.
https://github.com/googleapis/python-storage/blob/main/google/cloud/storage/blob.py#L1255

File: streaming/base/storage/upload.py (hunk 14 above)
Comment (karan6181) [reply streaming/base/storage/upload.py]: How about putting [this](https://github.com/mosaicml/streaming/blob/main/streaming/base/stream.py#L191) way ?

Comment (b-chu) [reply streaming/base/storage/upload.py]: Done.

**Review by karan6181 — APPROVED** [review]
LGTM. Thank You!
